src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -307,158 +379,101 @@ public override IAsyncResult BeginGetResponse(AsyncCallback callback, object sta
 
-            if (!headerSet && SettingContentLengthOnHeadersCollectionValidOnCurrentPlatform)
-            {
-                // Attempting to set the header but may just not work
-                try
-                {
-                    // Need this code to set the user agent on silverlight when the property doesn't exist
-                    httpWebRequest.Headers[XmlConstants.HttpContentLength] = XmlConvert.ToString(contentLength);
-                }
-                catch (ArgumentException)
+                // Apply cached "Content" header values
+                foreach (var contentHeader in _contentHeaderValueCache)
                 {
-                    SettingContentLengthOnHeadersCollectionValidOnCurrentPlatform = false;
+                    _requestMessage.Content.Headers.Add(contentHeader.Key, contentHeader.Value);
                 }
             }
-#else
-            httpWebRequest.ContentLength = contentLength;
-#endif
+
+            _requestMessage.Method = new HttpMethod(_effectiveHttpMethod);
+
+            var send = _client.SendAsync(_requestMessage);
+            return send;
         }
 
-        /// <summary>
-        /// Sets the accept charset.
-        /// </summary>
-        /// <param name="httpWebRequest">The HTTP web request.</param>
-        /// <param name="headerValue">The header value.</param>
-        internal static void SetAcceptCharset(HttpWebRequest httpWebRequest, string headerValue)
+        private static IDictionary<string, string> HttpHeadersToStringDictionary(HttpHeaders headers)
         {
-#if !PORTABLELIB
-            httpWebRequest.Headers[XmlConstants.HttpAcceptCharset] = headerValue;
-#endif
-#if PORTABLELIB
-            if (!IsRunningOnSilverlight)
-            {
-                httpWebRequest.Headers[XmlConstants.HttpAcceptCharset] = headerValue;
-            }
-#endif
+            return headers.ToDictionary((h1) => h1.Key, (h2) => string.Join(",", h2.Value));
         }
 
-        /// <summary>
-        /// Attempts to set the UserAgent property if it exists other wise returns false
-        /// </summary>
-        /// <param name="httpWebRequest">The http web request.</param>
-        /// <param name="headerValue">The value of the user agent.</param>
-        internal static void SetUserAgentHeader(HttpWebRequest httpWebRequest, string headerValue)
+        /// <summary> Returns the underlying HttpRequestMessage </summary>
+        public HttpRequestMessage HttpRequestMessage
         {
-#if !PORTABLELIB
-            httpWebRequest.UserAgent = headerValue;
-#endif
-#if PORTABLELIB
-            if (IsRunningOnSilverlight || !SettingUserAgentPropertyValidOnCurrentPlatform)
+            get
             {
-                return;
+                return _requestMessage;
             }
+        }
 
-            bool headerSet = false;
+        private static HttpWebResponseMessage ConvertHttpClientResponse(HttpResponseMessage response)

```

Comment:
Not sure I understand the naming for this method, especially the `HttpClient` part since internally it doesn't appear to reference anything related to `HttpClient`. Maybe `ConvertToHttpWebResponseMessage`?